### ISSUE TITLE
As a site visitor, if I have a WAAD login, I want to be able to use that username and password in CKAN.

### ISSUE THREAD

**Issue by adamamyl** [okfn/ckanext-glasgow#12]
As a platform user, if I have a Windows Azure Active Directory login, I want to be able to use that username and password in CKAN so I don't have another password to use.

The WAAD services will return an OAuth2 token (MS US R8.1.1: "An OAuth 2.0 token will be returned which will have to be used for all subsequent access to the platform")

When I login to CKAN, I expect CKAN to cache (until my next login) various informations from the corp-net SQL Azure database/API about me:
- [ ] the organization that account is associated with;
- [ ] the level I have in that organization (which is coterminous to the CKAN organization permissions);
- [ ] my email address;
- [ ] my full name

which will be used to populate my profile in CKAN, and the relevant organizations.
## External Dependencies

This user story _requires_ the WAAD, SQL Azure database, login resource and API to be built by Microsoft to function.
## Accept criteria
- [ ] Using a "GCC EC" valid username and password combination, a user can login to CKAN.
- [ ] Invalid user credentials will result in a denial of access to edit data on the platform
- [ ] As a part of the login process, CKAN should pull ancillary information stored in the supplementary database of Identity Services, and, where appropriate, render this in the user's profile.
- [ ] The user's profile details are cached on login, from EC -> CKAN store.

**Comment by adamamyl** [comment]
No need to factor in Single Sign OUT.

**Comment by adamamyl** [comment]
SLP to confirm what EC is doing.

**Comment by adamamyl** [comment]
@seanh Would you be interested in this? Perhaps testing out https://github.com/etalab/ckanext-oauth2 ? (documentation / endpoints for this is imminent)

**Comment by seanh** [comment]
Happy to look into it. Does developing this require a Windows Azure Active Directory account though? Sounds like a pain. Or does it just require a generic OAuth2 plugin?

**Comment by adamamyl** [comment]
We'll know more on the set-up today/tomorrow, but there will be a WAAD provisioned, and an account against which we will be able to test (for now, all accounts from WAAD will be admins, but that's changing).

In discussion today, we'll need to add the CKAN instance as a provider (in the Azure setup &#8212; details in lastpass), which can only take place once #24 has happened.

**Comment by seanh** [comment]
It seems we need to provision those Azure machines and get that WAAD and test account before we can start making any steps on this.

From what I read of the [OAuth 2.0 spec](http://tools.ietf.org/html/rfc6749) it sounds like OAuth 2 is not the kind of spec that you can write a generic client implementation of that will work against any OAuth 2 authorization and resource server, rather you need to implement a new client for each server specifically:

>   OAuth 2.0 provides a rich authorization framework with well-defined
> security properties.  However, as a rich and highly extensible
> framework with many optional components, on its own, this
> specification is likely to produce a wide range of non-interoperable
> implementations.
> 
> In addition, this specification leaves a few required components
> partially or fully undefined (e.g., client registration,
> authorization server capabilities, endpoint discovery).  Without
> these components, clients must be manually and specifically
> configured against a specific authorization server and resource
> server in order to interoperate.
> 
> This framework was designed with the clear expectation that future
> work will define prescriptive profiles and extensions necessary to
> achieve full web-scale interoperability.

So I think we need an instance of the actual OAuth 2 authorization and resource servers that we'll be working with, or complete documentation of their APIs, before we can start work.

It may be possible to have a generic ckanext-oauth2 extension that acts as a sort of "framework" making it easier to implement OAuth 2 plugins for different servers, much like how we write different harvester plugins for the harvester extension.

**Comment by seanh** [comment]
Looking to see if WAAD itself has any public docs about its implementation of OAuth 2.0...

**Comment by seanh** [comment]
... Nope, searching for that doesn't seem like an effective use of my time. Need documentation about the WAAD OAuth 2 service and a test server before we can start. Blocked for now I think

**Comment by adamamyl** [comment]
I've asked for a clarification/doc link from MS.

**Comment by adamamyl** [comment]
@seanh http://msdn.microsoft.com/en-us/library/azure/dn645545.aspx has been suggested.

**Comment by seanh** [comment]
http://msdn.microsoft.com/en-us/library/azure/dn645542.aspx combined with a test account for login.windows.net and with our CKAN instance added to Azure as a provider, should be enough to let someone login to CKAN using their WAAD username and password.

Once we register our CKAN instance with Azure and get a client ID, will I be able to use that client ID to do OAuth authorizations with their server from a development CKAN instance on my laptop? I hope so, otherwise I may have to do the development from one of our servers.

So once we have those I can get started developing that.

But I still need to know how to get those user details (organization, level, email address, full name) from the "corp-net SQL Azure database/API" so I can save them in CKAN. What is this API and are there some docs for it?

P.S. I love the use of "coterminous to" instead of "the same as"

**Comment by seanh** [comment]
The WAAD authentication server doesn't return any info about what organization the user is associated with, their level in that org, or their email address. @adamamyl do we have a separate ticket for that part?

I need clarification about this:

> As a part of the login process, CKAN should pull ancillary information stored in C, and, where appropriate, render this in the user's profile.

What is the supplementary database of Identity Services? And what ancillary information should we pull from it?

Apart from those two points, I think this ticket is done.

**Comment by adamamyl** [comment]
> The WAAD authentication server doesn't return any info about what organization the user is associated with, their level in that org, or their email address. @adamamyl do we have a separate ticket for that part?

All info that's not username/password is via the Identity Services database/API -- there's not a separate ticket per se.

In terms of ancillary info, I think it makes sense to grab:
- [ ] UserID (which should be unique)
- [ ] Name (concatenate both fields from the EC's first/last name fields) (= 'name' in ckan's user DB table?)
- [ ] Email address
- [ ] About blurb
- [ ] Associated Organization (if applicable)
- [ ] Role in associated org (org-admin, org-editor, org-member)

Display Name is set in the EC -- I don't know if we need it or not; I would think 'name' is in the CKAN code in lots of places… (but then again we have 'fullname' as a field in ckan's user db table).

-- Pretty much the stuff we display in the user profile.

The EC doesn't have the option for a gravatar, but I think we should keep that functionality in CKAN.

**Comment by seanh** [comment]
This isn't finished yet, login works but as I understand it we still need to grab the user's organization, role, email address and other info from the Services API.

**Comment by seanh** [comment]
Closing this, moved remaining requirements into this ticket: https://github.com/okfn/ckanext-glasgow/issues/82
